### DISCUSSION TITLE
Beginner Q: How to modify CC sent by Xone 1D as standalone hardware controller

### DISCUSSION THREAD

**Discussion by brianlion91** [RomanKubiak/ctrlr#558] [Q&A] [answered]
I want to use my Xone 1D midi hardware to control a Lexicon MX200 reverb processor (or similar) for my live DJ performance.

However, the transmitted midi cc of the Xone 1D midi hardware does not match the recognized midi cc of the Lexicon MX200.

Is there a way I can use Ctrlr to modify the midi map of the Xone 1D such that I can use it as a _standalone_ midi hardware controller for the Lexicon MX200? Any alternate suggestions??

Thanks for all your help!

[MX200_Manual_5073493-B.pdf](https://github.com/RomanKubiak/ctrlr/files/11658131/MX200_Manual_5073493-B.pdf)
[XONE1D_UG_AP7080_3.pdf](https://github.com/RomanKubiak/ctrlr/files/11658132/XONE1D_UG_AP7080_3.pdf)

**Comment by dnaldoog** [comment]
Can you give some more detail? For example could you show what code you are trying to use to do this? Or even post the panel on the forum?

**Reply by brianlion91** [reply]
Thanks for your reply. 

I am a total novice and although I've watched a few video tutorials on Ctrlr, I have not yet been able to determine whether this is the right software for my application. 

Hence why I asked if it's even possible to use Ctrlr to modify a CC midi map that might otherwise be hard-coded into an older piece of midi hardware (is it?). 

In other words, I want to use Ctrlr in a capacity similar to the Native Instruments Controller Editor application.

As such I haven't prepared any code or panels yet, though I look forward to sharing my work with the community once I get my bearings.

**Reply by damiensellier** [reply, marked as answer]
Absolutely, Ctrlr would be great for that. I started working in this kind of panel to translate Midi CC to another one.
The first thing you need to do is to list all the destination CC controls you have to use.
Then prepare a panel GUI with those knobs. 
You need to think if you want a GUI related to the hardware/software you want to control (ie your lexicon m200) or if you want to represent the remote control (xone 1D)
If it's the latter, design a panel with the different knobs as on the Xone. Assign the knobs to their original CC controller. 
Something based on that  diagram : 
![Xone 1D GUI](https://github.com/RomanKubiak/ctrlr/assets/55164385/f10c2423-8494-49c4-84fa-d3c2f768429a)

then you need lua to translate the incoming fixed parameter from the xone, to the new cc parameter value.
Under each knob on the panel you need a drop down (uiCombo) with the CC index 0-127. You chose your destination CC index with this and the lua script will translate the original value (the fixed one from the xone) to the new CC index from the drop-down.
Et voilà, you have a XONE cc translator application.

If you want a panel dedicated to the lexicon m200 that's another story. It would be the other way around.

Please confirm if this is what you want and I'll help you with that.
Damien

**Reply by damiensellier** [reply]
The following demo panel is for 1 knob, you can chose the midi channel destination and the CC number.

<img width="265" alt="CC Translator" src="https://github.com/RomanKubiak/ctrlr/assets/55164385/10d8c014-e0fc-440d-b4e0-8d48c463f22c">


[Basic CC Translator v0.1](https://github.com/RomanKubiak/ctrlr/files/11673504/CC.Translator_0_1_0_2023-06-07_02-54.panel.zip)

**Reply by brianlion91** [reply]
Hi Damien, thank you for your helpful replies. 

I totally get the concept and I'm confident that I can do most of this myself since I have experience with similar panel design in TouchOSC. Hopefully you can help me decide on the following two sticking points.

- Will it be realistic to run my Ctrlr CC translator panel on a dedicated Raspberry Pi?

- What configuration, operating system etc is best to automatically initialize my Ctrlr CC translator panel when I power up the Rasberry Pi?

The goal is to be able to remotely control the Lexicon MX200 with the Xone 1D so that it _feels_ like a computer-free experience while I'm DJing. At my gigs (for example at Burning Man later this summer), if more effort will be required than simply plugging in a Raspberry Pi to automatically run my Ctrlr panel, I might get impatient.

Perhaps I will need to dig in to the Raspberry Pi forums for the best way to implement this. 

Thank you again for your time and consideration.

Brian

**Reply by damiensellier** [reply]
Hi Brian, sorry but I never used a Raspberry Pi.
You should take a look at the midi library for arduino it's even less expensive. 
If you just need to translate midi CCs from the XONE1D to the MX200 it should be very easy with the arduino.
You just need to get the midi input message, get the byte of the midi message that defines the CC number, and make a lookup table to transform the message to your lexicon on the fly from your script.  Worst case scenario, just make a list of "case" in your midi CC translate function and for each case enter the target cc index of the lexicon.
  You don't even need a GUI, plug and play.

**Reply by brianlion91** [reply]
Arduino Midi Library sounds like it would be a simpler and more realistic for my minimalist requirements than Ctrlr.

And the Arduino board might even be small enough to install inside of the Xone 1D enclosure!

I am very grateful for your thoughtful and knowledgable responses -- I had never heard of Arduino Midi Library until just now.

I'll be hanging out in the github discussion for Midi Library. Thanks again Damien, it's been very nice chatting with you.

Sincerely,

Brian

**Reply by damiensellier** [reply]
You're welcome Brian, I discovered the Arduino midi capabilities thanks to "Look Mum No Computer" on youtube. Everytime he needs a midi controller he uses the arduino with a little bit of code. 

https://www.youtube.com/watch?v=wY1SRehZ9hM

I think your script would only take a few hours to get working. Midi In, translator, Midi Out and you're ready to rock the stage ;)
If ever you need to build a VST/AU panel for your lexicon give us a shout, the small CTRLR community is the best!
Damien

**Reply by brianlion91** [reply]
I'm sure it will be a worthwhile experience to learn to code in the Arduino development environment, but apparently ChatGPT can actually do this if get helplessly lost 
https://www.youtube.com/watch?v=dnrnOscJY2E&list=PLDvka7NV8VPB3jpWJ5NQe36J9nLdb8WFN&index=4

**Reply by damiensellier** [reply]
Oh yeah, I'm watching the video you posted now, and the AI is really good! The project is more complex than what you want, chatgpt will do it for you in a minute obviously.
